src/overlay/SurveyManager.cpp

```diff
@@ -0,0 +1,548 @@
+    topologyBody.totalOutboundPeerCount =
+        static_cast<uint32_t>(randomOutboundPeers.size());
+
+    try
+    {
+        response.encryptedBody = curve25519Encrypt<EncryptedBody::max_size()>(
+            request.encryptionKey, xdr::xdr_to_opaque(body));
+    }
+    catch (std::exception const& e)
+    {
+        CLOG(ERROR, "Overlay") << "curve25519Encrypt failed: " << e.what();
+        return;
+    }
+
+    auto sigBody = xdr::xdr_to_opaque(response);
+    signedResponse.responseSignature = mApp.getConfig().NODE_SEED.sign(sigBody);
+
+    broadcast(newMsg);
+}
+
+void
+SurveyManager::broadcast(StellarMessage const& msg) const
+{
+    mApp.getOverlayManager().broadcastMessage(msg, false,
+                                              MIN_OVERLAY_VERSION_FOR_SURVEY);
+}
+
+void
+SurveyManager::populatePeerStats(std::vector<Peer::pointer> const& peers,
+                                 PeerStatList& results,
+                                 VirtualClock::time_point now) const
+{
+    size_t resultSize = std::min<size_t>(peers.size(), results.max_size());
+    results.reserve(resultSize);
+
+    for (size_t i = 0; i < resultSize; ++i)
+    {
+        auto& peer = peers[i];
+        auto& peerMetrics = peer->getPeerMetrics();
+
+        PeerStats stats;
+        stats.id = peer->getPeerID();
+        stats.versionStr = peer->getRemoteVersion();
+        stats.messagesRead = peerMetrics.mMessageRead;
+        stats.messagesWritten = peerMetrics.mMessageWrite;
+        stats.bytesRead = peerMetrics.mByteRead;
+        stats.bytesWritten = peerMetrics.mByteWrite;
+
+        stats.uniqueFloodBytesRecv = peerMetrics.mUniqueFloodBytesRecv;
+        stats.duplicateFloodBytesRecv = peerMetrics.mDuplicateFloodBytesRecv;
+        stats.uniqueFetchBytesRecv = peerMetrics.mUniqueFetchBytesRecv;
+        stats.duplicateFetchBytesRecv = peerMetrics.mDuplicateFetchBytesRecv;
+
+        stats.uniqueFloodMessageRecv = peerMetrics.mUniqueFloodMessageRecv;
+        stats.duplicateFloodMessageRecv =
+            peerMetrics.mDuplicateFloodMessageRecv;
+        stats.uniqueFetchMessageRecv = peerMetrics.mUniqueFetchMessageRecv;
+        stats.duplicateFetchMessageRecv =
+            peerMetrics.mDuplicateFetchMessageRecv;
+
+        stats.secondsConnected =
+            std::chrono::duration_cast<std::chrono::seconds>(
+                now - peerMetrics.mConnectedTime)
+                .count();
+
+        results.emplace_back(stats);
+    }
+}
+
+void
+SurveyManager::recordResults(Json::Value& jsonResultList,
+                             PeerStatList const& peerList) const
+{
+    for (auto const& peer : peerList)
+    {
+        Json::Value peerInfo;
+        peerInfo["nodeId"] = KeyUtils::toStrKey(peer.id);
+        peerInfo["version"] = peer.versionStr;
+        peerInfo["messagesRead"] = static_cast<Json::UInt64>(peer.messagesRead);
+        peerInfo["messagesWritten"] =
+            static_cast<Json::UInt64>(peer.messagesWritten);
```

Comment:
@MonsieurNicolas I had to add these casts to `Json::UInt64` to fix the build.

Reply:
Funny that for Json they want uint64 and not int64 (js doesn’t deal with unsigned int 64 at all)